enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,737 @@
+
+[0] changing kernel arguments on ostree system results in creating new deployment.
+
+Example blueprint:
+
+```toml
+name = "microshift-low-latency"
+version = "0.0.1"
+modules = []
+groups = []
+distro = "rhel-94"
+
+[[packages]]
+name = "microshift"
+version = "4.17.*"
+
+[[packages]]
+name = "microshift-low-latency"
+version = "4.17.*"
+
+[[customizations.services]]
+enabled = ["microshift", "microshift-tuned"]
+
+[[customizations.kernel]]
+append = "some already known kernel args"
+name = "KERNEL-rt"
+
+[[customizations.files]]
+path = "/etc/tuned/microshift-baseline-variables.conf"
+data = """
+isolated_cores=1-2
+hugepagesDefaultSize = 2M
+hugepages2M = 128
+hugepages1G = 0
+additionalArgs = ""
+"""
+
+[[customizations.files]]
+path = "/etc/microshift/config.yaml"
+data = """
+kubelet:
+  cpuManagerPolicy: static
+  memoryManagerPolicy: Static
+"""
+
+[[customizations.files]]
+path = "/etc/microshift/microshift-tuned.yaml"
+data = """
+reboot_after_apply: True
+profile: microshift-baseline
+"""
+```
+
+
+##### bootc
+
+1. User creates Containerfile that:
+   - (optional) installs `kernel-rt`
+   - installs `microshift-low-latency.rpm`
+   - enables `microshift-tuned.service`
+   - adds following configs
+     - `/etc/tuned/microshift-baseline-variables.conf`
+     - `/etc/microshift/config.yaml` to configure Kubelet
+     - `/etc/microshift/microshift-tuned.yaml` to configure `microshift-tuned.service`
+1. User builds the blueprint
+1. User deploys the commit / installs the system.
+1. System boots - rest is just like in OSTree flow
+
+Example Containerfile:
+
+```
+FROM registry.redhat.io/rhel9/rhel-bootc:9.4
+
+# ... MicroShift installation ...
+
+RUN dnf install kernel-rt microshift-tuned
+COPY microshift-baseline-variables.conf /etc/tuned/microshift-low-latency-variables.conf
+COPY microshift-config.yaml             /etc/microshift/config.yaml
+COPY microshift-tuned.yaml              /etc/microshift/microshift-tuned.yaml
+
+RUN systemctl enable microshift-tuned.service
+```
+
+##### RPM
+
+1. User installs `microshift-low-latency` RPM.
+1. User creates following configs:
+   - `/etc/tuned/microshift-baseline-variables.conf`
+   - `/etc/microshift/config.yaml` to configure Kubelet
+1. Development environment
+   - User runs `sudo tuned-adm profile microshift-baseline` to enable the profile.
+   - User reboots the host to make changes to kernel arguments active.
+   - Host boots again, everything for low latency is in place,
+   - User starts/enables `microshift.service`
+1. Production environment
+   - User creates `/etc/microshift/microshift-tuned.yaml` to configure `microshift-tuned.service`
+   - User enables `microshift.service`
+   - User enables and starts `microshift-tuned.service` which:
+     - Compares active profile with requested profile
+     - If requested profile is already active:
+       - Compare checksum of requested profile with cached checksum.
+       - If checksums are the same - exit.
```

Comment:
What happens if there is an error here and the profile can't be applied? Does the previous profile apply (or none)?

Reply:
If `microshift-tuned.service` cannot reach step where it runs `tuned-adm profile PROFILE-NAME` (which activates the profile), then the old profile continues to be active.
If `tuned-adm profile PROFILE-NAME` fails (because e.g. profile references a variables.conf file that doesn't exist), then no profile could be active after such failure (because it cleans up old profile first, then applies new one)

Reply:
Since other people reading the docs might have the same question that Shauna asks, it might be a good idea to include this answer in the doc.

Reply:
I moved these bullet points to separate section "microshift-tuned service" where I explain the workflow once with greater detail